assets/src/edit-story/components/panels/layer/useLayerSelection.js

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { useStory } from '../../../app';
+
+function useLayerSelection( element ) {
+	const { type } = element;
+
+	const isBackground = type === 'background';
+	const backgroundHasElement = Boolean( element.inner );
+
+	const {
+		state: {
+			currentPage,
+			selectedElementIds,
+		},
+		actions: {
+			setSelectedElementsById,
+			toggleElementInSelection,
+			clearSelection,
+		},
+	} = useStory();
+
+	let isSelected;
+	if ( isBackground ) {
+		isSelected = backgroundHasElement ? selectedElementIds.includes( element.inner.id ) : selectedElementIds.length === 0;
+	} else {
+		isSelected = selectedElementIds.includes( element.id );
+	}
+
+	const pageElementIds = currentPage.elements.map( ( { id } ) => id );
+
+	const handleClick = useCallback( ( evt ) => {
+		const hasSelection = selectedElementIds.length > 0;
+
+		evt.preventDefault();
+		evt.stopPropagation();
+		if ( isBackground ) {
+			// If background layer is clicked either select nothing or select only background element
+			if ( backgroundHasElement ) {
+				setSelectedElementsById( { elementIds: [ element.inner.id ] } );
+			} else {
+				clearSelection();
+			}
+		} else if ( evt.shiftKey && hasSelection ) {
+			// Shift key pressed with any element selected:
+			// select everything between this layer and the first selected layer
+			const firstId = selectedElementIds[ 0 ];
+			const firstIndex = pageElementIds.findIndex( ( id ) => id === firstId );
+			const clickedIndex = pageElementIds.findIndex( ( id ) => id === element.id );
+			const lowerIndex = Math.min( firstIndex, clickedIndex );
+			const higherIndex = Math.max( firstIndex, clickedIndex );
+			const elementIds = pageElementIds.slice( lowerIndex, higherIndex + 1 );
+			// reverse selection if firstId isn't first anymore
+			if ( firstId !== elementIds[ 0 ] ) {
+				elementIds.reverse();
+			}
+			setSelectedElementsById( { elementIds } );
+		} else if ( evt.metaKey ) {
+			// Meta pressed. Toggle this layer in the selection.
+			toggleElementInSelection( { elementId: element.id } );
+		} else {
+			// No special key pressed - just selected this layer and nothing else.
+			setSelectedElementsById( { elementIds: [ element.id ] } );
+		}
```

Comment:
So here I went for the "classic"_*_ list selection mode of:

* clicking something without keyboard modifier to select it and only it,
* meta-clicking something to toggle it in selection, or
* shift-clicking something to select everything between **it** and the **first selection** made. This will preserve the first selection.

So, an example of this is as follows:

* click layer 5 (selection is now: `[5]`)
* then shift-click layer 6 (selection is now: `[5,6]`)
* then shift-click layer 3 (selection is now: `[5,4,3]`, note reverse order)
* then shift-click layer 7 (selection is now: `[5,6,7]`)
* then meta-click layer 6 (selection is now: `[5,7]`)
* then meta-click layer 3 (selection is now: `[5,7,3]`)
* then meta-click layer 5 (selection is now: `[7,3]`)
* then shift-click layer 2 (selection is now: `[7,6,5,4,3,2]`, note that 7 is the new "first" element)
* then shift-click layer 7 (selection is now: `[7]`)

This will require some fairly complex testing IMHO.

* I think this follows classic selection patterns. No?

Reply:
I see this as something to be generalized and re-used for selection for pages in the grid view for #58 perhaps. Thoughts, @swissspidy?

Reply:
Indeed that sounds like a good thing re-use in other components like the grid view as well. Or perhaps even the carousel. The Figma design for those doesn't really mention that though.

Reply:
I had a chat with Sam and @dvoytenko about it, and we agreed to keep it like this for now and see how it is received in user testing.